Detectors/ZDC/reconstruction/src/DigiReco.cxx

```diff
@@ -0,0 +1,785 @@
+    if (fval < 0) {
+      // Check if calibration object is present
+      if (!mTDCParam) {
+        LOG(FATAL) << "TDC " << itdc << " missing configuration object and no manual override";
+      } else {
+        fval = mTDCParam->getShift(itdc) / FTDCVal;
+      }
+    }
+    auto val = std::nearbyint(fval);
+    if (val < kMinShort) {
+      LOG(FATAL) << "Shift for TDC " << itdc << " " << val << " is out of range";
+    }
+    if (val > kMaxShort) {
+      LOG(FATAL) << "Shift for TDC " << itdc << " " << val << " is out of range";
+    }
+    tdc_shift[itdc] = val;
+    LOG(INFO) << itdc << " " << ChannelNames[TDCSignal[itdc]] << " shift= " << tdc_shift[itdc] << " i.s. = " << val * FTDCVal << " ns";
+  }
+  // TDC search zone
+  // TODO: override with configuration object
+  for (int itdc = 0; itdc < o2::zdc::NTDCChannels; itdc++) {
+    LOG(INFO) << itdc << " " << ChannelNames[TDCSignal[itdc]] << " search= " << ropt.tdc_search[itdc] << " i.s. = " << ropt.tdc_search[itdc] * FTDCVal << " ns";
+  }
+
+  // Fill maps channel maps for integration
+  for (int ich = 0; ich < NChannels; ich++) {
+    // If the reconstruction parameters were not manually set
+    if (ropt.amod[ich] < 0 || ropt.ach[ich] < 0) {
+      for (int im = 0; im < NModules; im++) {
+        for (uint32_t ic = 0; ic < NChPerModule; ic++) {
+          if (mModuleConfig->modules[im].channelID[ic] == ich && mModuleConfig->modules[im].readChannel[ic]) {
+            ropt.amod[ich] = im;
+            ropt.ach[ich] = ic;
+            goto next_ich;
+          }
+        }
+      }
+    }
+  next_ich:;
+    LOG(INFO) << "ADC " << ich << "(" << ChannelNames[ich] << ") mod " << ropt.amod[ich] << " ch " << ropt.ach[ich];
+  }
+
+  // Integration ranges
+  for (int ich = 0; ich < NChannels; ich++) {
+    // If the reconstruction parameters were not manually set
+    if (ropt.beg_int[ich] <= DummyIntRange || ropt.end_int[ich] <= DummyIntRange) {
+      if (!mIntParam) {
+        LOG(ERROR) << "Integration for signal " << ich << " missing configuration object and no manual override";
+      } else {
+        ropt.beg_int[ich] = mIntParam->beg_int[ich];
+        ropt.end_int[ich] = mIntParam->end_int[ich];
+      }
+    }
+    if (ropt.beg_ped_int[ich] <= DummyIntRange || ropt.end_ped_int[ich] <= DummyIntRange) {
+      if (!mIntParam) {
+        LOG(ERROR) << "Integration for pedestal " << ich << " missing configuration object and no manual override";
+      } else {
+        ropt.beg_ped_int[ich] = mIntParam->beg_ped_int[ich];
+        ropt.end_ped_int[ich] = mIntParam->end_ped_int[ich];
+      }
+    }
+    LOG(INFO) << ChannelNames[ich] << " integration: signal=[" << ropt.beg_int[ich] << ":" << ropt.end_int[ich] << "] pedestal=[" << ropt.beg_ped_int[ich] << ":" << ropt.end_ped_int[ich] << "]";
+  }
+}
+
+int DigiReco::process(const std::vector<o2::zdc::OrbitData>* orbitdata, const std::vector<o2::zdc::BCData>* bcdata, const std::vector<o2::zdc::ChannelData>* chdata)
```

Comment:
I assume this method is called from the DigitRecoSpec, passing the spans extracted from the inputs?
Then here also one should use e.g. `gsl::span<const o2::zdc::OrbitData> orbitData` etc.

Reply:
Changed method to use span instead of vectors